+ added nimsuggest support for exception inlay hints

nickysn opened the pull request:
This adds nimsuggest support for displaying inlay hints for exceptions. An inlay hint is displayed around function calls, that can raise an exception, which isn't handled in the current subroutine (in other words, exceptions that can propagate back to the caller). On mouse hover on top of the hint, a list of exceptions that could propagate is shown.

The changes, required to support this are already commited to nimlangserver and the VS code extension. The extension and the server allow configuration for whether these new exception hints are enabled (they can be enabled or disabled independently from the type hints), as well as the inlay strings that are inserted before and after the name of the function, around the function call. Potentially, one of these strings can be empty, for example, the user can choose to add an inlay hint only before the name of the function, or only after the name of the function.

nickysn committed:
+ added support for exception inlay hints

nickysn committed:
* refactored the nimsuggest symbol database to allow sorting and binary searches

nickysn committed:
* nimsuggest tester: in case of parseInt exception, when parsing the port number, echo the actual data we got from the nimsuggest process for diagnostic reasons

nickysn committed:
* fixed syntax errors in previous commit

nickysn committed:
+ nimsuggest tester: print the entire nimsuggest output for diagnostic purposes in case of parseInt error

nickysn committed:
* fixed error: field 'sym' is not accessible for type 'TNode' using 'kind = nkIdent' [FieldDefect]

Araq commented on compiler/modulegraphs.nim:
```diff
@@ -55,10 +55,12 @@ type
     inst*: PInstantiation
 
-  SymInfoPair* = object
+  SymInfoPair* = ref object
```
Does it really need to be `ref`? The compiler suffers from a massive overuse of `ref` making things harder to understand than necessary.

nickysn replied:
How do we set caughtExceptions then? This is done in a different pass, after the SymInfoPair objects have been collected. Then, we need to update them, by setting caughtExceptions. If it's not a ref object, we should copy the entire object, change a single field, then copy it again? Isn't that worse than using a 'ref object'? Or, is there an easier way?

Araq replied:
Why not update it in-place or `move` the object?

nickysn replied:
Ok: https://github.com/nim-lang/Nim/pull/23202/commits/9ff6cd988a2e1351e1f4ef5d0fa6ad7ba15ee935

nickysn committed:
* SymInfoPair type converted back to a non-ref object

nickysn committed:
Merge branch 'devel' into try_inlay_hint5

Araq commented on nimsuggest/tester.nim:
```diff
@@ -279,7 +279,13 @@ proc runEpcTest(filename: string): int =
+    except ValueError:
+      echo "Error parsing int: " & a
+      echo outp.readAll()
+      raise
```
What does that do? Improve the error message but does not handle the error (it uses re-raising)? But who had problems with the previous error message?

nickysn replied:
During the development of this feature, I had a nimsuggest crash that only happened on the GitHub CI machines, which I could not reproduce on my computer. Therefore, I added this, so that the stack trace, produced by nimsuggest is printed on the screen, and becomes visible in the GitHub CI log. The bug, that caused the crash has been fixed, but I left this here, because it's useful for diagnostic purposes, even though it's unrelated to the inlay hints for exceptions. I can extract it to a separate PR if you want?

Araq replied:
Ideally you simply remove it as I don't like it. If you want a better/different error mesage handle the error completely aka replace the `raise` by `quit`.

nickysn replied:
Done: https://github.com/nim-lang/Nim/pull/23202/commits/a7bf527cde473f1f7be17ebe3ac17a38c948e199

Araq commented on compiler/modulegraphs.nim:
```diff
@@ -712,12 +726,44 @@ func belongsToStdlib*(graph: ModuleGraph, sym: PSym): bool =
+proc add*(s: var SuggestSymbolDatabase; v: SymInfoPair) =
+  s.mgetOrPut(v.info.fileIndex, newSuggestFileSymbolDatabase()).add(v)
+
+proc findSymInfo*(s: var SuggestFileSymbolDatabase; li: TLineInfo): ref SymInfoPair =
```
Why not use `ptr SymInfoPair`?

nickysn replied:
I saw that other functions in nimsuggest.nim return `ref SymInfoPair` - `findSymData`, `findByTLineInfo`, so I used `ref` for consistency. Is using `ptr` the preferred way? Should the other functions that return `ref SymInfoPair` also be changed to `ptr SymInfoPair`?

Araq replied:
If possible, yes. We don't wrap types with `ref` just so that we can get `nil` as a possible value. We use `ptr` for that. Of course, you need to watch out things are still memory safe but it's usually the case.

nickysn replied:
Changed the result to `ptr` in `findSymInfo`: https://github.com/nim-lang/Nim/pull/23202/commits/6a22c4a2ca6bb672dc440b90b47acb8112b19c21
Should I change the others here, or should that be in a separate PR?

Araq commented on compiler/modulegraphs.nim:
```diff
@@ -712,12 +726,44 @@ func belongsToStdlib*(graph: ModuleGraph, sym: PSym): bool =
+  var idx = binarySearch(s.items, q, cmp)
+  if idx != -1:
+    new(result)
+    result[] = s.items[idx]
```
Why not use `result = addr s.items[idx]`?

nickysn replied:
Done: https://github.com/nim-lang/Nim/commit/6a22c4a2ca6bb672dc440b90b47acb8112b19c21

Araq commented:
How does this affect the memory consumption of nimsuggest?

nickysn commented:
> How does this affect the memory consumption of nimsuggest?

I hope the memory use won't increase much, but this needs to be measured. I'll try to do some measurements and report what I found.

nickysn committed:
* return `ptr SymInfoPair` instead of `ref SymInfoPair` in findSymInfo

nickysn committed:
* nimsuggest tester: in case of error parsing the port number, handle the error completely and quit, instead of reraising the exception

nickysn committed:
Merge branch 'devel' into try_inlay_hint5

nickysn commented:
> > How does this affect the memory consumption of nimsuggest?
> 
> I hope the memory use won't increase much, but this needs to be measured. I'll try to do some measurements and report what I found.

Ok, I measured the memory use of nimsuggest on two files - nimsuggest.nim (an example of a large-sized project, as it includes the Nim compiler) and nimlangserver.nim (a medium-sized project). I used `smem -k` to measure the memory use. I'm running the tests on Fedora 39 for x86_64. I compared nimsuggest from this feature branch with nimsuggest from the devel branch, after synchronizing this branch with the devel branch. The results:

nimsuggest.nim:
before:
  PID User     Command                         Swap      USS      PSS      RSS 
42451 nickysn  nimsuggest /home/nickysn/tr        0   550.8M   550.8M   552.8M 
after:
  PID User     Command                         Swap      USS      PSS      RSS 
45030 nickysn  nimsuggest /home/nickysn/tr        0   591.8M   591.8M   593.7M 

That's about 7.4% increase.

nimlangserver.nim:
before:
  PID User     Command                         Swap      USS      PSS      RSS 
43033 nickysn  nimsuggest /home/nickysn/wo        0   219.4M   219.4M   221.4M 
after:
  PID User     Command                         Swap      USS      PSS      RSS 
45588 nickysn  nimsuggest /home/nickysn/wo        0   278.5M   278.5M   280.4M 

That's about 26% increase.

I'm a little surprised that we get a larger increase (percentage-wise) on smaller projects, compared to large projects. I would expect it to be the opposite. Maybe it's an artifact of the measurement method or the way the heap is managed (how it grows)?

Araq commented:
Can be this made opt-in somehow? Not everybody enjoys a nimsuggest that uses more memory just so that it reports exception tracking details which many don't care about.

nickysn committed:
* reordered fields in SymInfoPair to reduce memory use. No functional changes.

nickysn commented:
> Can be this made opt-in somehow? Not everybody enjoys a nimsuggest that uses more memory just so that it reports exception tracking details which many don't care about.

I'm thinking about this. Meanwhile I've commited a small optimization that makes the memory increase slightly less. Now it's:
for nimsuggest.nim:
286965 nickysn  nimsuggest /home/nickysn/tr        0   590.4M   590.4M   592.5M 
That's a 7.2% increase (previously, it was 7.4%).

for nimlangserver.nim:
287981 nickysn  nimsuggest /home/nickysn/wo        0   256.9M   258.4M   262.0M 
That's a 17% increase (previously, it was 26%).

It's still not opt-in. I'm still thinking about how to optimize this SymInfoPair structure. Since it's so critical, there are a lot of optimizations possible if it's redesigned, however these changes would probably be unrelated to the exception hints.

nickysn commented:
One way to make this opt-in is to move the caught exceptions data to a parallel array (in other words, to partially move from an array of structs to a structure of arrays). This way, if exception hints are turned off, we can avoid allocating the extra memory. If there are no better suggestions, and if I don't come with a better solution, I'll try to implement this approach.

Araq commented:
Ok, let's see how this outline solution of yours looks like.

nickysn committed:
* refactored the SuggestFileSymbolDatabase structure, so that it contains
  caughtExceptions as a separate array. This will allow for the memory for
  caughtExceptions not to be allocated, when the exception inlay hints are not
  enabled (which is not done, yet). Many things are refactored to allow
  converting SuggestFileSymbolDatabase from an array of structs, to a struct of
  arrays. Extracting further fields as separate arrays will result in further
  reduction of used memory, even when inlay hints for exceptions are turned off,
  since the SymInfoPair structure wastes too much memory, due to alightment
  padding.

nickysn committed:
* extracted caughtExceptionsSet as a separate array, in order to reduce wasted
  memory, due to padding (this will actually happen in the next commit, when
  isDecl is moved as well).
* fixed the commented out setting of caughtExceptionsSet that was temporarily
  disabled in the previous commit

nickysn committed:
* fixed findSymInfoIndex not returning result. Oops.

nickysn committed:
* extracted isDecl as a separate array in SuggestFileSymbolDatabase

nickysn committed:
* extracted sym as a separate array in SuggestFileSymbolDatabase

nickysn committed:
* replaced TLineInfo in SuggestFileSymbolDatabase with just the pair line, col, without FileIndex in order to save memory.

nickysn committed:
* InternalSymInfoPair renamed TinyLineInfo, as that's a more appropriate name,
  after the refactoring. No functional changes.

nickysn committed:
* SuggestFileSymbolDatabase.items renamed SuggestFileSymbolDatabase.lineInfo.
  No functional changes.

nickysn committed:
+ added nimsuggest command line option `--exceptionInlayHints:on|off` that
  allows to globally turn the exception inlay hints on or off. They are now off
  by default. The extra memory for the tracking of caught exceptions is only
  allocated, when these inlay hints are turned on.

nickysn committed:
* fixed the parsing of the new `--exceptionInlayHints:on|off` nimsuggest command line switch

nickysn commented:
Added optimizations and made the feature possible to turn on or off via command line switch. When turned off, memory use is even less than before. New results with the exception hints enabled:

With exception hints enabled:
 for nimsuggest.nim:
104793 nickysn  nimsuggest /home/nickysn/tr        0   554.7M   556.3M   559.8M 
around 1% increase

for nimlangserver.nim:
104259 nickysn  nimsuggest /home/nickysn/wo        0   241.2M   242.7M   246.2M 
around 10% increase

With exception hints disabled:
for nimsuggest.nim:
106228 nickysn  nimsuggest /home/nickysn/tr        0   541.9M   543.4M   547.0M 
about 1.6% decrease

for nimlangserver.nim:
106561 nickysn  nimsuggest /home/nickysn/wo        0   219.0M   220.6M   224.1M 
about 0.5%-1.2% decrease

Now, time to resolve the merge conflicts...

nickysn committed:
Merge branch 'devel' into try_inlay_hint5

nickysn committed:
* fixed off-by-one bug, accidentally introduced in commit 866e1e62583e8ed1d8b36b5ff5690a32e1451b0f

beef331 commented on compiler/modulegraphs.nim:
```diff
@@ -74,6 +80,18 @@ type
+    lineInfo*: seq[TinyLineInfo]
+    sym*: seq[PSym]
+    caughtExceptions*: seq[seq[PType]]
+    caughtExceptionsSet*: seq[bool]
```
For memory concerns should this be `caughtExceptionsSet*: IntSet`? Where the index of `caughtExceptions` is stored inside.

nickysn replied:
Yes, same for isDecl. This will further reduce the memory used. However, I was planning to do this in a separate PR, after this one is merged, because this PR is already large enough.

nickysn replied:
Well, since this PR is being ignored for such a long time, I've decided to not wait for it to be merged, but to go ahead and implement the IntSet optimization as well. Here it is: https://github.com/nim-lang/Nim/pull/23202/commits/5ff0f229a390f24837576ca6d8c28889c72c5b16

nickysn committed:
Merge branch 'devel' into try_inlay_hint5

nickysn committed:
* fixed the parsing of multiple extra parameters to the 'inlayhint' command. For
  forward and backwards compatibility, ignore unknown parameters.

nickysn committed:
* fixed exception inlay hints for calls via procvars

nickysn committed:
* exception inlay hints: assume the worst (i.e. raises Exception) if a procvar
  has no explicit raises pragma

nickysn committed:
+ also show inlay exception hints for calls via procvar function parameters, but
  don't show them if the parameter is marked '.effectsOf'

nickysn committed:
+ properly handle tokens, starting with a backtick at the given column (versus
  the previous column) inside getTokenLenFromSource

nickysn committed:
Merge branch 'devel' into try_inlay_hint5

nickysn committed:
Merge branch 'devel' into try_inlay_hint5

nickysn committed:
Merge branch 'devel' into try_inlay_hint5

nickysn committed:
Merge branch 'devel' into try_inlay_hint5

nickysn committed:
Merge branch 'devel' into try_inlay_hint5

nickysn committed:
Merge branch 'devel' into try_inlay_hint5

nickysn committed:
Merge branch 'devel' into try_inlay_hint5

nickysn committed:
+ introduced PackedBoolArray, a bitpacked array, based on IntSet. Use that,
  instead of seq[bool] in SuggestFileSymbolDatabase.caughtExceptionsSet and
  SuggestFileSymbolDatabase.isDecl, in order to save memory.

Araq commented:
Move `SuggestSymbolDatabase` and its related procs into its own module.

nickysn committed:
Moved SuggestSymbolDatabase and its related procs to a new module: suggestsymdb.nim

nickysn commented:
> Move `SuggestSymbolDatabase` and its related procs into its own module.

Done: https://github.com/nim-lang/Nim/pull/23202/commits/00f6d55f1862a4fc60f2cde470cb086659384cd4

SirOlaf commented:
This seems to have some issues inferring exceptions of generic procs.
```nim
proc foo[T](x: T) =
  discard

foo(1)
```
A bell for the call foo(1) appears, supposedly raising "Exception".
~~I am unsure if this a new issue or if this simply exposes an existing one.~~
Looks like exceptions from generic procs didn't propagate properly before either, but this makes it obvious by putting `Exception` and inserting bells everywhere.

Removing the `getEbase` call in initialization here https://github.com/nim-lang/Nim/pull/23202/files#diff-a7c8d25cefdf0d3b4d7713ada23b1ec76416139cdaaa9999c0e8ef8fda94cb2aR931 gets rid of some false positives but then it simply thinks generics never raise.

Issue is that the symbol table holds the generic variant instead of an instance, so propagation still works as it happens in sempass2, but the immediate ide hint is complete nonsense.

arnetheduck commented:
> Looks like exceptions from generic procs didn't propagate properly before either

I guess that's the conservative choice - you can't know which exceptions will be raised until you instantiate it - `when` is the most trivial way (but consider calling an overloaded function from `foo`): 
```nim
proc foo[T](x: T) =
  when T is int: raise ValueError 
  else: raise OSError

foo(1)
```

SirOlaf commented:
It propagates without issues in the sempass2 step of compiler, nimsuggest does not get the instance symbol and that is what fails. I tried hacking together a fix to demonstrate this in #23414
Should work normally if that is fixed properly.
